File: longestIncreasingPath.py
from collections import deque

class Solution:

    def kahnAlgorithm(self, n, m, adj, indegree):
        order = []
        q = deque()

        for i in range(n * m):
            if indegree[i] == 0:
                q.append(i)

        while q:
            node = q.popleft()
            order.append(node)

            for neighbor in adj[node]:
                indegree[neighbor] -= 1
                if indegree[neighbor] == 0:
                    q.append(neighbor)

        return order

    def longestIncreasingPath(self, matrix):
        n = len(matrix)
        m = len(matrix[0])

        adj = [[] for _ in range(n * m)]
        indegree = [0] * (n * m)

        dir = [0, -1, 0, 1, 0]

        for i in range(n):
            for j in range(m):
                for k in range(4):
                    x = i + dir[k]
                    y = j + dir[k + 1]

                    if 0 <= x < n and 0 <= y < m and matrix[x][y] > matrix[i][j]:
                        adj[i * m + j].append(x * m + y)
                        indegree[x * m + y] += 1

        order = self.kahnAlgorithm(n, m, adj, indegree)

        dp = [1] * (n * m)

        for it in order:
            for edge in adj[it]:
                dp[edge] = max(dp[edge], dp[it] + 1)

        return max(dp)

File: longestIncreasingPath.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class Solution {

    public List<Integer> kahnAlgorithm(int n, int m, List<Integer>[] adj, int[] indegree) {
        List<Integer> order = new ArrayList<>();
        Queue<Integer> q = new LinkedList<>();

        for (int i = 0; i < n * m; i++) {
            if (indegree[i] == 0) {
                q.add(i);
            }
        }

        while (!q.isEmpty()) {
            int node = q.poll();
            order.add(node);

            for (int neighbor : adj[node]) {
                indegree[neighbor]--;
                if (indegree[neighbor] == 0) {
                    q.add(neighbor);
                }
            }
        }

        return order;
    }

    public int longestIncreasingPath(int[][] matrix) {
        int n = matrix.length;
        int m = matrix[0].length;

        List<Integer>[] adj = new ArrayList[n * m];
        int[] indegree = new int[n * m];

        int[] dir = {0, -1, 0, 1, 0};

        for (int i = 0; i < n * m; i++) {
            adj[i] = new ArrayList<>();
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                for (int k = 0; k < 4; k++) {
                    int x = i + dir[k];
                    int y = j + dir[k + 1];

                    if (x >= 0 && y >= 0 && x < n && y < m && matrix[x][y] > matrix[i][j]) {
                        adj[i * m + j].add(x * m + y);
                        indegree[x * m + y]++;
                    }
                }
            }
        }

        List<Integer> order = kahnAlgorithm(n, m, adj, indegree);

        int[] dp = new int[n * m];
        Arrays.fill(dp, 1);

        for (int it : order) {
            for (int edge : adj[it]) {
                dp[edge] = Math.max(dp[edge], dp[it] + 1);
            }
        }

        return Arrays.stream(dp).max().getAsInt();
    }
}
